ring/lifecycler.go

```diff
@@ -0,0 +1,885 @@
+	NumTokens                int           `yaml:"num_tokens"`
+	HeartbeatPeriod          time.Duration `yaml:"heartbeat_period"`
+	ObservePeriod            time.Duration `yaml:"observe_period"`
+	JoinAfter                time.Duration `yaml:"join_after"`
+	MinReadyDuration         time.Duration `yaml:"min_ready_duration"`
+	InfNames                 []string      `yaml:"interface_names"`
+	FinalSleep               time.Duration `yaml:"final_sleep"`
+	TokensFilePath           string        `yaml:"tokens_file_path"`
+	Zone                     string        `yaml:"availability_zone"`
+	UnregisterOnShutdown     bool          `yaml:"unregister_on_shutdown"`
+	ReadinessCheckRingHealth bool          `yaml:"readiness_check_ring_health"`
+
+	// For testing, you can override the address and ID of this ingester
+	Addr string `yaml:"address" doc:"hidden"`
+	Port int    `doc:"hidden"`
+	ID   string `doc:"hidden"`
+
+	// Injected internally
+	ListenPort int `yaml:"-"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *LifecyclerConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.RegisterFlagsWithPrefix("", f)
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet.
+func (cfg *LifecyclerConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	cfg.RingConfig.RegisterFlagsWithPrefix(prefix, f)
+
+	// In order to keep backwards compatibility all of these need to be prefixed
+	// with "ingester."
+	if prefix == "" {
+		prefix = "ingester."
+	}
+
+	f.IntVar(&cfg.NumTokens, prefix+"num-tokens", 128, "Number of tokens for each ingester.")
+	f.DurationVar(&cfg.HeartbeatPeriod, prefix+"heartbeat-period", 5*time.Second, "Period at which to heartbeat to consul. 0 = disabled.")
+	f.DurationVar(&cfg.JoinAfter, prefix+"join-after", 0*time.Second, "Period to wait for a claim from another member; will join automatically after this.")
+	f.DurationVar(&cfg.ObservePeriod, prefix+"observe-period", 0*time.Second, "Observe tokens after generating to resolve collisions. Useful when using gossiping ring.")
+	f.DurationVar(&cfg.MinReadyDuration, prefix+"min-ready-duration", 15*time.Second, "Minimum duration to wait after the internal readiness checks have passed but before succeeding the readiness endpoint. This is used to slowdown deployment controllers (eg. Kubernetes) after an instance is ready and before they proceed with a rolling update, to give the rest of the cluster instances enough time to receive ring updates.")
+	f.DurationVar(&cfg.FinalSleep, prefix+"final-sleep", 30*time.Second, "Duration to sleep for before exiting, to ensure metrics are scraped.")
+	f.StringVar(&cfg.TokensFilePath, prefix+"tokens-file-path", "", "File path where tokens are stored. If empty, tokens are not stored at shutdown and restored at startup.")
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		level.Error(log.NewNopLogger()).Log("msg", "failed to get hostname", "err", err)
```

Comment:
This was originally using the (util_logger) global logger, but now we don't wish to do that. I could add a logger to the config, but I'm not sure if that's the best path here. Right now I just create a new no-op logger which preserves the line, but doesn't log. Is there a system logger or a generic logger I could use in go? Or could I return an error and have that be uncaught since we send an os exit after this log?

Reply:
I think you're right, we definitely don't want `os.Exit` in library code. Returning an error is the right choice here.

Reply:
On the other hand, a panic feels fine in the case we can't look up the hostname of the machine this code is one. That seems like something that Should Never Fail.

Reply:
I took @56quarters suggestion and used a panic here.